projects/iio_adpd1080/src/main.c

```diff
@@ -47,13 +47,168 @@
+	struct irq_init_param cal_irq_init = {
+		.irq_ctrl_id = 0,
+		.platform_ops = &aducm_irq_ops,
+		.extra = NULL
+	};
+
+	status = timer_init(&cal_timer, &cal_timer_init);
+	if(status != SUCCESS)
+		return FAILURE;
+	status = timer_start(cal_timer);
+	if(status != SUCCESS)
+		goto timer_finish;
+
+	status = gpio_get(&sync_gpio, &sync_gpio_init);
+	if(status != SUCCESS)
+		goto timer_finish;
+	status = gpio_direction_input(sync_gpio);
+	if(status != SUCCESS)
+		goto gpio_finish;
+
+	status = irq_ctrl_init(&cal_irq, &cal_irq_init);
+	if(status != SUCCESS)
+		goto gpio_finish;
+
+	struct gpio_irq_config sync_irq_config = {
+		.gpio_handler = sync_gpio,
+		.mode = GPIO_GROUP_POSITIVE_EDGE
+	};
+	struct callback_desc sync_gpio_cb = {
+		.callback = adpd1080_sync_gpio_cb,
+		.ctx = &sync_gpio_pulse_no,
+		.config = &sync_irq_config
+	};
+	status = irq_register_callback(cal_irq, ADUCM_GPIO_A_INT_ID,
+				       &sync_gpio_cb);
+	if(status != SUCCESS)
+		goto gpio_finish;
+
+	status = adpd188_adc_fsample_set(adpd1080_dev, 1000);
+	if(status != SUCCESS)
+		goto finish;
+
+	status = adpd188_reg_write(adpd1080_dev, ADPD188_REG_GPIO_DRV,
+				   ADPD188_GPIO_DRV_GPIO0_ENA_MASK);
+	if(status != SUCCESS)
+		goto finish;
+
+	status = adpd188_reg_write(adpd1080_dev, ADPD188_REG_GPIO_CTRL, 2);
+	if(status != SUCCESS)
+		goto finish;
+
+	status = adpd188_mode_set(adpd1080_dev, ADPD188_NORMAL);
+	if (status != SUCCESS)
+		goto finish;
+
+	status = irq_enable(cal_irq, ADUCM_GPIO_A_INT_ID);
+	if(status != SUCCESS)
+		goto finish;
+
+	while (true) {
+		status = adpd188_reg_read(adpd1080_dev, ADPD188_REG_SAMPLE_CLK, &temp_reg);
+		if(status != SUCCESS)
+			goto finish;
+		status = timer_counter_get(cal_timer, &t_start);
+		if(status != SUCCESS)
+			goto finish;
+		sync_gpio_pulse_no = 0;
+		while (t_start >= t_stop) {
+			status = timer_counter_get(cal_timer, &t_stop);
+			if(status != SUCCESS)
+				goto finish;
+		}
+		if (sync_gpio_pulse_no < 1000)
+			temp_reg -= 1;
+		else if (sync_gpio_pulse_no > 1000)
+			temp_reg += 1;
+		if (abs(sync_gpio_pulse_no - 1000) >= min_diff)
+			break;
+		else
+			min_diff = abs(sync_gpio_pulse_no - 1000);
+		status = adpd188_reg_write(adpd1080_dev, ADPD188_REG_SAMPLE_CLK, temp_reg);
+		if(status != SUCCESS)
+			goto finish;
+	}
+
+	status = irq_disable(cal_irq, ADUCM_GPIO_A_INT_ID);
+	if(status != SUCCESS)
+		goto finish;
+
+	status = irq_unregister(cal_irq, ADUCM_GPIO_A_INT_ID);
+	if(status != SUCCESS)
+		goto gpio_finish;
+
+	status = adpd188_mode_set(adpd1080_dev, ADPD188_PROGRAM);
+	if (status != SUCCESS)
+		goto finish;
+
+	status = adpd188_adc_fsample_set(adpd1080_dev, 16);
+	if(status != SUCCESS)
+		goto finish;
+
+	status = adpd188_reg_write(adpd1080_dev, ADPD188_REG_GPIO_DRV, 0);
+	if(status != SUCCESS)
+		goto finish;
+
+	status = adpd188_reg_write(adpd1080_dev, ADPD188_REG_GPIO_CTRL, 0);
+	if(status != SUCCESS)
+		goto finish;
+
+finish:
+	fin_ret = irq_ctrl_remove(cal_irq);
+	if(fin_ret != SUCCESS)
+		return FAILURE;
```

Comment:
if you return here, `cal_timer` and `sync_gpio` memory might remain allocated.

Reply:
If freeing the memory inside cal_irq's remove() function is failing, probably freeing the memory reserved for the gpio won't work either.

Reply:
I think it wouldn't be wrong to try :)